### PR TITLE
Interactive create

### PR DESCRIPTION
**Pull request by SmileyChris**
# Description
If no filename is given when doing `towncrier create`, interactively ask for the issue number and fragment type (and then launch an interactive editor for the newsfragment content).

```console
$ towncrier create
Issue number (`+` for none): 482
Fragment type (feature, bugfix, doc, removal, misc): feature
($EDITOR opened here)
Created news fragment at /home/chris/dev/towncrier/src/towncrier/newsfragments/482.feature.rst
```

## Reworded interactive edit comment
The comment given to the editor has been reworded to use the similar format as a git commit interactive edit, with the empty space above rather than below:
```
█
# Please write your news content. Lines starting with '#' will be ignored, and
# an empty message aborts.
```

## Default news fragment filename extension
If the file extension of the `filename` configuration setting ends in `.rst` or `.md`, `towncrier create` will append this extension to newsfragments it creates (if the filename didn't provide an explicit extension).
```console
$ towncrier create 1.feature
Created news fragment at /home/chris/dev/towncrier/src/towncrier/newsfragments/1.feature.rst
$ towncrier create 2.feature.txt
Created news fragment at /home/chris/dev/towncrier/src/towncrier/newsfragments/2.feature.txt
```
This is configurable with a new `create_add_extension` setting (`true` by default).

## Add an empty line to the end of news fragments
An empty line is now added to the end of news fragment content, whether entered by the command line option or interactively.
This is configurable with a new `create_eof_newline` setting (`true` by default).

# Checklist

* [x] Make sure changes are covered by existing or new tests.
* [x] For at least one Python version, make sure local test run is green.
* [x] Create a file in `src/towncrier/newsfragments/`. Describe your
  change and include important information. Your change will be included in the public release notes.
* [x] Make sure all GitHub Actions checks are green (they are automatically checking all of the above).
* [x] Ensure `docs/tutorial.rst` is still up-to-date.
* [x] If you add new **CLI arguments** (or change the meaning of existing ones), make sure `docs/cli.rst` reflects those changes.
* [x] If you add new **configuration options** (or change the meaning of existing ones), make sure `docs/configuration.rst` reflects those changes.

### CODE REVIEW

**Comment by adiroiban** [comment]
Thanks Chris for this nice PR.
I will try to find some time to review this...but I am very busy this need and next week.
I hope someone else will have time to review it.
If someone left unreviewed don't hesitate to ping me and I will to review.

At a first review, my initial comment was to move the configuration changes into a separate PR and focus this PR only on the interactive part.
In this way, it might be easier to merge one part of this PR, without waiting for other part that might need more work/feedback.

But I am also ok to review this together. It's just that it will take more time to review.

Thanks again for your help.

**Comment by SmileyChris** [comment]
> But I am also ok to review this together. It's just that it will take more time to review.

Yes, I threw in a bit more than I should have... but it was all intertwined and it was too messy to provide them as individual PRs. If any area becomes contentious, I'm happy to remove it from this PR.

Preemptively defending the secondary changes:
* **Interactive edit rewording**: this should be pretty straight forward, it just brings things into line with a git interactive commit, and having the empty line first is a nice advantage to allow typing without navigation in most editors.
* **Default news fragment**: this is a completely optional addition, but one that fit nicely with recent discussions on file extensions we were having. Having a `.rst` (or `.md`) extension makes the experience nicer for users who have an editor that styles content based on file type.
* **Adding empty lines to fragments**: functionally it doesn't change anything with generated output, but it is good practice (and without, fragments technically not even technically text files according to IEEE Std 1003.1-2001: A [text file](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_205) consists of a series of [lines](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap03.html#tag_03_205), each of which requires a terminating newline character). A common pre-commit hook adds a newline the end of files.

**Comment by adiroiban** [comment]
Thanks. I will try to take a look at this next week.

**Comment by adiroiban** [comment]
Sorry Chris for this late review.
It was still marked as unread... but went at the bottom of my github notifications list.

If you are still interested in this. I think that as long as the tests are green, this can be merged.

I don't know whey I will have time to test this.

File: src/towncrier/create.py
Comment (adiroiban) [inline src/towncrier/create.py:38]: Why use the `None` placeholder here ?

File: src/towncrier/create.py
Comment (adiroiban) [inline src/towncrier/create.py:54]: Why not leave `edit` as always being boolean ?

It looks like we are using `None` here as a placeholder for the default value.

But maybe it's better to have the explicit default value defined in this way.

```suggestion
    edit: bool = True,
```

File: src/towncrier/create.py
Comment (adiroiban) [inline src/towncrier/create.py]: Can we have the default value defined here?

```suggestion
    default=False,
```

File: src/towncrier/newsfragments/482.feature.rst
Comment (adiroiban) [inline src/towncrier/newsfragments/482.feature.rst:1]: I think that the `--add-extension` is also a new feaure introduced by this PR.

Also, `--eof-neline` is a new feature of this PR.

Should't we have release notes for these features ?

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py:124]: Maybe add a comment that the conten is pass `verbatim` and no extra newline or processed is done.

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py]: `Ensure` is kind of redundand here.
We write tests to make sure all works as expected :)

```suggestion
    def test_no_filename_extension(self, runner: CliRunner):
        """
        When the NEWS filename has no extensions,
        new fragments are created by default without any
        extension.
        """
```

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py]:
```suggestion
        # The new of the file where towncrier will generate
        # the final release notes is names `RELEASE_NOTES`
        # and for this test, has no file extension.
        setup_simple_project(extra_config='filename = "RELEASE_NOTES"')
```

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py]: This looks like the same docstring from `        """Ensure we don't overwrite existing files."""`

I think that this needs a better docstring to explain this test scenarion.

What we are setting up here, what we are doing and what we are expecting.

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py:413]: It looks like the expectation was changed here, without any changes in the test setup.

Is this a change that modified the default behaviour ?

I think that is ok to change the default behaviour, but this needs to be mentioned in the release notes.

This project is all about release notes, so I think it's important for `towncrier` to be a good example of how the release notes should be managed :)

File: src/towncrier/test/test_create.py
Comment (adiroiban) [inline src/towncrier/test/test_create.py:443]: Again a change that looks like it's modifying the default behaviour.

**Review by adiroiban — CHANGES_REQUESTED** [review]
Sorry for the late review. and many thanks for the PR. it was easy to follow.

I did a quick code review.

I didn't had time to get the code and run manual tests.

Please see my inline comments and let me know if they make sense.

I think the changes are ok, but it's important to have better release notes for the changes here.

it looks like besides interactive create, this PR also adds 2 new features and breaks at least 1 backward compatibility :)

We can merge this as it is, as long as we document the changes as part of the release notes.

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py]: No, since the default of `None` allows falling back to using the configuration value. Setting `True` or `False` allows a user to always override that configuration default.

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py:54]: This is only set to `True` only if the content is still the default content. If a user specifies their own content then, we don't want to default to triggering editing mode. But by leaving it `None`, it makes it possible to provide custom content **and** still explicitly choose to enable editing mode with this flag.

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py:38]: Discussed below

File: src/towncrier/test/test_create.py
Comment (SmileyChris) [reply src/towncrier/test/test_create.py:413]: Now mentioned in release notes

File: src/towncrier/test/test_create.py
Comment (SmileyChris) [reply src/towncrier/test/test_create.py:443]: Now mentioned in release notes

File: src/towncrier/test/test_create.py
Comment (SmileyChris) [reply src/towncrier/test/test_create.py:124]: Because that comment would be incorrect. The newline option is still processed after it's received back from the editor.

Comment (SmileyChris) [reply src/towncrier/test/test_create.py:124]: This is specifically testing the case mentioned elsewhere when a user provides custom content but still chooses to edit their content first.

File: src/towncrier/create.py
Comment (adiroiban) [reply src/towncrier/create.py:54]: We can have it like this... it looks like we are hijacking this variable :)

We should have a separate flag to signal the other use case.

File: src/towncrier/_settings/load.py
Comment (adiroiban) [inline src/towncrier/_settings/load.py:56]: Should these 2 new config options have the default set to False, for backward compatibility ?

File: src/towncrier/create.py
Comment (adiroiban) [inline src/towncrier/create.py]: Just trying to understand what is going on here 

```suggestion
    if eof_newline is None:
        # If no explicit flag was provided via CLI
        # default to the value from the configuration file.
        eof_newline = config.create_eof_newline
```

File: src/towncrier/create.py
Comment (adiroiban) [reply src/towncrier/create.py]: Do we need this as a command line option?

Can we just go with the config file ?

File: src/towncrier/test/test_create.py
Comment (adiroiban) [reply src/towncrier/test/test_create.py:124]: OMG :) Are people using this feature ?

I am ok with that, we can have it.

It looks so complicated :) ... but I have never felt the need for `towncrier create`

For me it's always easier to just create the file directly in my text editor, without calling towncrier.

File: src/towncrier/newsfragments/482.feature.rst
Comment (adiroiban) [inline src/towncrier/newsfragments/482.feature.rst]: I am ok with this new default value... but maybe it should be placed into a separate section ... as it looks like a backward incompatible change.

File: src/towncrier/newsfragments/482.feature.rst
Comment (adiroiban) [inline src/towncrier/newsfragments/482.feature.rst:3]: Not sure if this is feature, or a backward compatiblity break :)

**Review by adiroiban — COMMENTED** [review]
Thanks for the update.

I don't have much time for this reaview...and I don't really understand the need for `towncrier create` ... in general :)

So, I am not the best person to review this.

But we should try go get this merged.

As long as it has docs + tests, it should be ok.

My comments are about:

* backward compatibilty , related to the default values of the new config options
* adding yet another CLI argument to `towncrier create` ... maybe having the value in config file is enough


I hope @hynek or @altendky can take a look and see if we need the extra CLI argument and it it's ok to introduce this backward compatiblity change.

Thanks again

--------

I am using a towncrier version from 2017 ...and I am happy with that. So I don't really care.

I am just trying to maintain this, as part of my release management work for twisted/twisted

File: src/towncrier/_settings/load.py
Comment (SmileyChris) [reply src/towncrier/_settings/load.py:56]: It depends if backwards compatibility is more important than (imo) more sensible defaults. I'll leave that decision to the maintainers to make a call on.

File: src/towncrier/newsfragments/482.feature.rst
Comment (SmileyChris) [reply src/towncrier/newsfragments/482.feature.rst:3]: It's a change in the filename but since the file is parsed by towncrier in exactly the same way, I don't think it's really backwards incompatible

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py]: Yeah, it's a pretty obscure setting to ever need to change.

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py:54]: It's the same behaviour as any of the other cli settings that also have a configuration setting. If None, fall back to the configuration, otherwise use the value as provided.

File: src/towncrier/create.py
Comment (SmileyChris) [reply src/towncrier/create.py]: Flag gone now

File: src/towncrier/newsfragments/482.feature.rst
Comment (SmileyChris) [reply src/towncrier/newsfragments/482.feature.rst]: As in a separate fragment? I'm not sure where else you mean this should go.

File: src/towncrier/_settings/load.py
Comment (adiroiban) [reply src/towncrier/_settings/load.py:56]: I am not using the create command... so I don't know what to say.

In general, the rule is to keep backward compatibiliy. Any exception should be explicitly raised and accepted on a case by case basis.

We don't have a separate backward compatiblity policy for towncrier ... so I go with the general Twisted org policy https://docs.twisted.org/en/stable/development/compatibility-policy.html

I am happy to have a separate policy for towncrier.
Since towncrie is a dev tool, I think we can be more releaxed.

At the same time, I don't like when a CI pipeline breaks after an update 

I guess that `towncrier create` is not used as part of any automation... so we should be ok.

If nobody else is -1 on this default, I will merge is at it is.

I am happy with the new defaults.

Thanks!

File: src/towncrier/newsfragments/482.feature.rst
Comment (adiroiban) [reply src/towncrier/newsfragments/482.feature.rst]: Usualy backward incopatible changes are announced into `482.removal.rst`  This is the "Deprecation and removal" category.

I know that "removal" is a bad name... but this is what we have.

There are a log of bad defaults and bad naming conventions in towncrier :(

bad as in non intuitive

File: src/towncrier/newsfragments/482.feature.rst
Comment (adiroiban) [reply src/towncrier/newsfragments/482.feature.rst:3]: OK. Thanks. Fine by me.

I am not sure that adding release notes in this way, in a single fragment, does what we want.

I remember that towncrier will create one list item per file

On my project, as a hack I have `482-1.feature.rst` as a way to add another item in the notes...but I think that we need better support here

I remember there were other disussions about this

**Review by adiroiban — APPROVED** [review]
Thanks.

All good.

The only think blocking this is moving the backward incomaptible change into a `.removal` fragment

I will wait a bit to see if anyone else has anything agains this backward incopatible change and all is ok, this can be merged.

thanks again!

**Comment by adiroiban** [comment]
@SmileyChris I think that you are already in the top 5 contributors of this repo for the last few years, so I consider you one of the de-facto maintainers 


I sent you an invite to a team that has write access to the repo.
This will get you write access and issue triage access. You should be able to add tags and update issues

**Comment by adiroiban** [comment]
I found a bit of time today, so I have triggered the auto-merge.

I am not sure when I will have time to check this.

If anyone is -1 on these changes, we can still revert them before the next public release.

Thanks for all your help!

**Pull request merged**
